popclient.man



popclient(1L)                                       popclient(1L)


NAME
       popclient  -  retrieve  mail  from a mailserver using Post
       Office Protocol.

SYNOPSIS
       popclient  [-2  |  -3]  [-Vksv]  [-u  server-userid]   [-p
       server-password]
       [-f remote-folder] [-c | -o local-folder]  host

DESCRIPTION
       popclient   is  a  Post  Office  Protocol  compliant  mail
       retrieval client which supports both POP2 (as specified in
       RFC 937) and POP3 (RFC 1225).

       Typically,  popclient  will  be  used  to download mail in
       batch from the remote mailserver specified by  host  to  a
       mail  folder  on  the local disk.  The retrieved mail will
       then be manipulated using a local  mail  reader,  such  as
       mail or elm.

       To  facilitate the use of popclient in scripts, pipelines,
       etc, it returns an appropriate exit code upon  termination
       -- see EXIT CODES below.

OPTIONS
       -2     Use Post Office Protocol version 2 (POP2).

       -3     Use Post Office Protocol version 3 (POP3).

       -k     Keep messages in folder on remote mailserver.  Nor-
              mally, messages are deleted from the folder on  the
              mailserver  after they have been retrieved.  Speci-
              fying -k causes retrieved  messages  to  remain  in
              your folder on the mailserver.

       -s     Silent  mode.   Suppresses all progress/status mes-
              sages that are normally echoed to stderr  during  a
              POP  connection.  If both the -s and -v options are
              specified, the -v option takes precedence.

       -v     Verbose mode.  All control messages passed  between
              popclient  and the mailserver are echoed to stderr.
              Specifying -v causes  normal  progress/status  mes-
              sages which would be redundant or meaningless to be
              modified or omitted.

       -u     Specifies the user idenfication  to  be  used  when
              logging-in to the mailserver.  The appropriate user
              identification is both server and  user  dependent.
              Default  is  your login name on the machine that is
              running popclient.  See USER  AUTHENTICATION  below
              for a complete description.

       -p     Specifies  the  password to be used when logging-in



                                                                1





popclient(1L)                                       popclient(1L)


              to the mailserver.   The  appropriate  password  is
              both  server  and user dependent.  If the -p option
              is not used to specify  a  password,  you  will  be
              prompted  for  a  password before the connection to
              the mailserver is established.  See USER  AUTHENTI-
              CATION below for a complete description.

       -f     Causes  an  alternate mail folder on the mailserver
              to be retrieved.  The syntax of the folder name  is
              server  dependent,  as is the default behavior when
              no folder  is  specified.   Fortunately,  most  POP
              servers  have a reasonable default behavior, so use
              of this option should be limited to fairly special-
              ized  applications.  POP3 does not provide a folder
              specification in the protocol.  If the -f option is
              used  in conjunction with the -3 option, the remote
              folder specification is ignored.

       -o     Causes retrieved messages  to  be  appended  to  an
              alternate mail folder on the local disk.  When nei-
              ther -o nor -c is specified, retrieved messages are
              appended  to  the  system  default mail folder. See
              OUTPUT OPTIONS below for a complete description.

       -c     Causes retrieved messages to be written  to  stdout
              instead of a mail folder.  See OUTPUT OPTIONS below
              for a complete description.  You  may  not  specify
              both  the  -c  and  -o  options on the same command
              line.

       -V     Displays the version information for your  copy  of
              popclient.  If you specify the -V option, all other
              options are ignored and no POP connection is  made.


PROTOCOL SELECTION
       The selection of the correct Post Office Protocol (POP2 or
       POP3) depends upon the  configuration  of  the  mailserver
       from  which you retrieve your mail.  The system adminstra-
       tor who installed popclient on  your  system  should  have
       chosen   an   appropriate   default   protocol   for  your
       mailserver.  If you get the message  'Connection  refused'
       when  using  the default protocol, try specifying -2 or -3
       to  select  a  different  protocol.   If  the  'Connection
       refused'  message  persists  regardless  of  the  protocol
       selected, it is likely that your mailserver is not running
       a POP compliant mail service.


USER AUTHENTICATION
       User  authentication  in  popclient  is very much like the
       authentication mechanism of ftp(1).  The  correct  user-id
       and password depend upon the underlying security system at
       the mailserver.



                                                                2





popclient(1L)                                       popclient(1L)


       If the mailserver is a Unix machine on which you  have  an
       ordinary  user  account, your regular login name and pass-
       word are used with popclient.  If you use the  same  login
       name  on  both  the  server  and  the client machines, you
       needn't worry about  specifying  a  user-id  with  the  -u
       option -- the default behavior will use your login name on
       the client machine as the user-id on the  server  machine.
       If  you  use a different login name on the server machine,
       specify that login name with the -u option.  e.g. if  your
       login name is 'jsmith' on a machine named 'mailgrunt', you
       would start popclient as follows:

              popclient -u jsmith mailgrunt

       The default behavior of popclient is  to  prompt  you  for
       your  mailserver  password  before  the  POP connection is
       established.  This is the safest way to use popclient  and
       ensures  that  your password will not be compromised.  You
       may also specify your password using the -p option.   This
       is convenient when using popclient with automated scripts,
       but it may result in your password being exposed to prying
       eyes  --  be  careful!  Regardless of how your password is
       specified it is never stored in shared memory segments, or
       left  unencrypted  in the core image when popclient termi-
       nates.  Continuing the  preceding  example,  suppose  your
       password  on 'mailgrunt' is 'Gr8PassWd'.  The syntax would
       be:

              popclient -u jsmith -p Gr8PassWd mailgrunt

       On mailservers that do not provide ordinary user accounts,
       your  user-id  and  password  are  usually assigned by the
       server administrator when you apply for a mailbox  on  the
       server.   Contact  your  server administrator if you don't
       know the correct user-id and  password  for  your  mailbox
       account.


OUTPUT OPTIONS
       popclient  always writes the retrieved messages using Unix
       mail folder format.  This allows popclient to be  used  in
       conjunction  with  common  mail readers like mail and elm.
       The retrieved  messages  are  normally  appended  to  your
       default  system mailbox on the local disk, using the local
       Mail Delivery Agent (MDA), usually /bin/mail(1),  so  that
       when  you  invoke  your  mail reader it can manipulate the
       retrieved messages like any other mail you receive on  the
       client machine.

       Using  the  -o  option,  you  can specify a different mail
       folder to which the retrieved messages will  be  appended.
       If  you  prefer, for example, to have your POP mail from a
       machine called 'mailgrunt' stored in the mbox file in your
       home directory, you would start popclient as follows:



                                                                3





popclient(1L)                                       popclient(1L)


              popclient -o $HOME/mbox mailgrunt

       Note  that  the  folder specified with -o is not locked or
       otherwise protected from other  processes  writing  to  it
       while popclient is writing to it.

       popclient  can be used in a shell pipeline by using the -c
       option.  In this mode, popclient writes the retrieved mes-
       sages  to  stdout,  instead  of a mail folder.  This would
       allow you, for instance, to pass the incoming mail through
       a  filter that discards mail marked as 'Precedence: junk'.
       Suppose you've written an AWK script called 'dumpjunk.awk'
       to  implement  a junk mail filter.  The appropriate syntax
       to retrieve your mail from 'mailgrunt',  pass  it  through
       the  filter, and write it to a folder called 'realmail' in
       your home directory would be:

              popclient -c mailgrunt  |  awk  -f  dumpjunk.awk  >
              $HOME/realmail

       The progress/status messages written to stderr when the -s
       option has not been specified, do not interfere  with  the
       message  stream, which is written to stdout.  You may even
       use -v and -c  together  without  corrupting  the  message
       stream.  It is a good idea to use the -k option when using
       -c to insure that your messages will not be lost  if  part
       of  the shell pipeline does not function incorrectly.  The
       safest bet would be something like:

              popclient   -k   -c   mailgrunt   |   myfilter    >
              $HOME/filtered.mail

       followed by

              popclient -c mailgrunt > /dev/null

       when  you're sure the messages were correctly processed by
       'myfilter'.


EXIT CODES
       To facilitate the use of popclient in  shell  scripts  and
       the  like,  an exit code is returned to give an indication
       of what occured during a given POP connection.   The  exit
       code  can  be  tested by the script and appropriate action
       taken.

       A simple example follows.  This Bourne shell  script  exe-
       cutes  popclient  and,  if some messages were successfully
       retrieved from a mailserver  retrieved  from  the  command
       line,  it  starts the mail utility to read those messages.
       Otherwise,  it  prints  a  brief   message,   and   exits.
       #!/bin/sh




                                                                4





popclient(1L)                                       popclient(1L)


       if popclient $1 then
         mail else
         echo "No mail to read."  fi

       The exit codes returned by popclient are as follows:

       0      One or more messages were successfully retrieved.

       1      There was no mail awaiting retrieval.

       2      An  error was encountered when attempting to open a
              socket for the POP connection.  If you  don't  know
              what  a  socket  is,  don't  worry about it -- just
              treat this as an 'unrecoverable error'.

       3      The user authentication step failed.  This  usually
              means that a bad user-id or password was specified.

       4      Some sort of protocol error was detected.   POP  is
              not  especially  forgiving  when  it comes to unex-
              pected responses, commands,  etc  --  the  protocol
              invariably  calls  for  terminating  the connection
              under such error conditions.

       5      There was a syntax error in the arguments  to  pop-
              client.

       6      Some  kind of I/O woes occurred when writing to the
              local folder.

       7      There was an error condition reported by the server
              (POP3 only).

       9      Something  totally undefined occured.  This is usu-
              ally caused by a bug within popclient.  Do  let  me
              know if this happens.


AUTHOR
       popclient was written by Carl Harris at Virginia Polytech-
       nic Institute and State University (a.k.a. Virginia Tech).


BUGS
       There  are  none!  Well, maybe one or two.  Send comments,
       bug reports, gripes, and the like to ceharris@vt.edu.

SEE ALSO
       mail(1), binmail(1), sendmail(8), popd(8),  RFC  937,  RFC
       1225.







                                                                5




source/popclient.c
/* Copyright 1993,1994 by Carl Harris, Jr.
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Carl Harris <ceharris@vt.edu>
 */

 
/***********************************************************************
  module:       popclient.c
  program:      popclient
  SCCS ID:      @(#)popclient.c	2.5  4/1/94
  programmer:   Carl Harris, ceharris@vt.edu
  date:         29 December 1993
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  main driver module for the popclient package.  contains
                code for command-line parsing and invocation of the
                appropriate POP driver routine.
 ***********************************************************************/
 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <pwd.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/file.h>
#include <sys/wait.h>
#include <sys/stat.h>

#include "config.h"
#include "popclient.h"

/* release info */
#define         COMPILE_SCCS	"2.21 [2.5] (4/1/94)"

/* requisite, venerable SCCS ID string */
static char sccs_id [] = "@(#)popclient.c	2.5\t4/1/94";

#ifndef NO_PROTO
/* prototypes for internal functions */
int showoptions (struct optrec *options);
int parseMDAargs (struct optrec *options);
int parsecmdline (int argc, char **argv, struct optrec *options);
int setdefaults (struct optrec *options);
int showversioninfo (void);
int clearbiffbit (void);
int restorebiffbit (void);
#endif

/* Controls the detail of status/progress messages written to stderr */
int outlevel;      /* see the O_.* constants in popclient.h */
int biffwas = -1;  /* previous chmod bits on the controlling tty */

/* args for the MDA, parsed out in the usual fashion by parseMDAargs() */
#ifdef MDA_ARGS
char *mda_argv [MDA_ARGCOUNT + 2];
#else
char *mda_argv [2];
#endif


/*********************************************************************
  function:      main
  description:   main driver routine 
  arguments:     
    argc         argument count as passed by runtime startup code.
    argv         argument strings as passed by runtime startup code.

  return value:  an exit status code for the shell -- see the 
                 PS_.* constants defined above.
  calls:         parsecmdline, setdefaults, openuserfolder, doPOP2.
  globals:       none.
 *********************************************************************/

main (argc,argv)
int argc;
char **argv;
{ 
  int mboxfd;
  struct optrec options;
  int popstatus;
  int cleanup();

  signal(SIGINT,cleanup);
  clearbiffbit();

  if (parsecmdline(argc,argv,&options) == 0)
    if (!options.versioninfo)
      if (setdefaults(&options) == 0) {
        parseMDAargs(&options);
        if (options.whichpop == 3) 
          popstatus = doPOP3(&options);
        else
          popstatus = doPOP2(&options);
      } 
      else
        popstatus = PS_UNDEFINED;
    else
      showversioninfo();
  else
    popstatus = PS_SYNTAX;

  restorebiffbit();
  exit(popstatus);
}
    

 
/*********************************************************************
  function:      parsecmdline
  description:   parse/validate the command line options.
  arguments:
    argc         argument count.
    argv         argument strings.
    options      pointer to a struct optrec to receive the parsed 
                 options.

  return value:  zero if the command line was syntactically correct,
                 else non-zero (an error message has been written to stderr).  
  calls:         none.  
  globals:       none.  
 *********************************************************************/

int parsecmdline (argc,argv,options)
int argc;
char **argv;
struct optrec *options;
{
  int c,i;
  int fflag = 0;     /* TRUE when -o or -c has been specified */
  int errflag = 0;   /* TRUE when a syntax error is detected */
  extern int optind,opterr;     /* defined in getopt(2) */
  extern char *optarg;          /* defined in getopt(2) */

  bzero(options,sizeof(struct optrec));    /* start clean */
  while (!errflag && (c = getopt(argc,argv,"23Vkvscu:p:f:o:")) != EOF) {
    switch (c) {
      case '2':
        options->whichpop = 2;
        break;
      case '3':
        options->whichpop = 3;
        break;
      case 'V':
        options->versioninfo = !0;
        break;
      case 'k':
        options->keep = !0;
        break;
      case 'v':
        options->verbose = !0;
        break;
      case 's':
        options->silent = !0;
        break;
      case 'c':
        if (fflag)
          errflag++;
        else {
          fflag++;
          options->foldertype = OF_STDOUT;
        }
        break;
      case 'u':
        strcpy(options->userid,optarg);
        break;
      case 'p':
        strcpy(options->password,optarg);
        break;
      case 'o':
        if (fflag) 
          errflag++;
        else {
          fflag++;
          options->foldertype = OF_USERMBOX;
          strcpy(options->userfolder,optarg);
        }
        break;
      case 'f':
        strcpy(options->remotefolder,optarg);
        break;
      default:
        errflag++;
    }
  }

  if (!options->versioninfo) {
    /* get host argument -- error if not present */
    if (optind < argc)
      strcpy(options->host,argv[optind++]);
    else
      errflag++;

    /* error if more unparsed arguments are present */
    if (optind < argc)
      errflag++;
    else
      ;
  }
  else 
    ;


  /* squawk if syntax errors were detected */
  if (errflag) {
    fputs("usage:  popclient [-2|-3] [-Vksv] [-u server-userid] [-p server-password]\n",
          stderr);
    fputs("                  [-f remote-folder] [-c | -o local-folder]  host\n",
          stderr);
  }
  else
    ;

  /* clean up the command line in case -p was used */
  for (i = 1; i < argc; i++)
    argv[i] = (char *) 0;

  return(errflag);
}
         

/*********************************************************************
  function:      setdefaults
  description:   set reasonable default values for unspecified options.
  arguments:     
    options      option values parsed from the command-line; unspeci-
                 fied options must be filled with zero.

  return value:  zero if defaults were successfully set, else non-zero
                 (indicates a problem reading /etc/passwd).
  calls:         none.
  globals:       writes outlevel.
 *********************************************************************/

int setdefaults (options)
struct optrec *options;
{
  int uid;
  struct passwd pw1, *pw = &pw1;
  char *mailvar;

/*
  if ((pw = getpwuid(uid = getuid())) == NULL) {
    fprintf(stderr,"No passwd entry for uid %d\n",uid);
    return(-1);
  }
*/
  pw->pw_name = (getenv("USER") ? getenv("USER") : "root");

  /* save the login name for delivery use */
  strcpy(options->loginid,pw->pw_name);

  if (options->whichpop == 0)
    options->whichpop = DEF_PROTOCOL;
  else
    ; /* -2 or -3 was specified */

  if (*(options->userid) == '\0') 
    strcpy(options->userid,pw->pw_name);
  else
    ;  /* -u was specified */

  if (options->foldertype == 0) 
    options->foldertype = OF_SYSMBOX;
  else
    ;  /* -o or -c was specified */
    
  if (*(options->password) == '\0')  
    strcpy(options->password,(char *) getpass("Enter mailserver password: "));
  else
    ;  /* -p was specified */

  if (options->verbose)
    outlevel = O_VERBOSE;
  else if (options->silent)
    outlevel = O_SILENT;
  else
    outlevel = O_NORMAL;

  return(0);
}



/*********************************************************************
  function:      showversioninfo
  description:   display program release and compiler info
  arguments:     none.
  return value:  none.
  calls:         none.
  globals:       none.
 *********************************************************************/

int showversioninfo()
{
  printf("popclient release %s\n",COMPILE_SCCS);
}



/*********************************************************************
  function:      openuserfolder
  description:   open the file to which the retrieved messages will
                 be appended.  Do NOT call when options->foldertype
                 is OF_SYSMBOX.

  arguments:     
    options      fully-determined options (i.e. parsed, defaults invoked,
                 etc).

  return value:  file descriptor for the open file, else -1.
  calls:         none.
  globals:       none.
 *********************************************************************/

int openuserfolder (options)
struct optrec *options;
{
  int fd;

  if (options->foldertype == OF_STDOUT)
    return(1);
  else    /* options->foldertype == OF_USERMBOX */
    if ((fd = open(options->userfolder,O_CREAT|O_WRONLY|O_APPEND,0600)) >= 0) {
      return(fd);
    }
    else {
      perror("popclient: openuserfolder: open()");
      return(-1);
    }
  
}



/*********************************************************************
  function:      openmailpipe
  description:   open a one-way pipe to the mail delivery agent.
  arguments:     
    options      fully-determined options (i.e. parsed, defaults invoked,
                 etc).

  return value:  open file descriptor for the pipe or -1.
  calls:         none.
  globals:       reads mda_argv.
 *********************************************************************/

int openmailpipe (options)
struct optrec *options;
{
  int pipefd [2];
  int childpid;
  char binmailargs [80];

  if (pipe(pipefd) < 0) {
    perror("popclient: openmailpipe: pipe");
    return(-1);
  }
  if ((childpid = fork()) < 0) {
    perror("popclient: openmailpipe: fork");
    return(-1);
  }
  else if (childpid == 0) {

    /* in child process space */
    close(pipefd[1]);  /* close the 'write' end of the pipe */
    close(0);          /* get rid of inherited stdin */
    if (dup(pipefd[0]) != 0) {
      fputs("popclient: openmailpipe: dup() failed\n",stderr);
      exit(1);
    }

    execv(MDA_PATH,mda_argv);

    /* if we got here, an error occurred */
    perror("popclient: openmailpipe: exec");
    return(-1);

  }

  /* in the parent process space */
  close(pipefd[0]);  /* close the 'read' end of the pipe */
  return(pipefd[1]);
}



/*********************************************************************
  function:      closeuserfolder
  description:   close the user-specified mail folder.
  arguments:     
    fd           mail folder descriptor.

  return value:  zero if success else -1.
  calls:         none.
  globals:       none.
 *********************************************************************/

int closeuserfolder(fd)
int fd;
{
  int err;

  if (fd != 1) {   /* not stdout */
    err = close(fd);
  }   
  else
    err = 0;
  
  if (err)
    perror("popclient: closeuserfolder: close");

  return(err);
}



/*********************************************************************
  function:      closemailpipe
  description:   close pipe to the mail delivery agent.
  arguments:     
    options      fully-determined options record
    fd           pipe descriptor.

  return value:  0 if success, else -1.
  calls:         none.
  globals:       none.
 *********************************************************************/

int closemailpipe (fd)
int fd;
{
  int err;
  int childpid;

  err = close(fd);
  childpid = wait((union wait *) 0);
  if (err)
    perror("popclient: closemailpipe: close");

  return(err);
}



/*********************************************************************
  function:      parseMDAargs
  description:   parse the argument string given in MDA_ARGS into
                 a regular *argv[] array.
  arguments:
    options      fully-determined options record pointer.

  return value:  none.
  calls:         none.
  globals:       writes mda_argv.
 *********************************************************************/

int parseMDAargs (options)
struct optrec *options;
{
  int argi;
  char *argp;

  /* first put the MDA alias in as argv[0] */
  mda_argv[0] = MDA_ALIAS;
  
#ifdef MDA_ARGS

  /* make a writeable copy of MDA_ARGS */
  argp = strcpy((char *) malloc(strlen(MDA_ARGS)+1), MDA_ARGS);
  
  while (*argp != '\0' && isspace(*argp))	/* skip null first arg */
    argp++;					

  /* now punch nulls into the delimiting whitespace in the args */
  for (argi = 1;  
       *argp != '\0';
       argi++) {

    mda_argv[argi] = argp;     /* store pointer to this argument */

    /* find end of this argument */
    while (!(*argp == '\0' || isspace(*argp)))
      argp++;

    /* punch in a null terminator */
    if (*argp != '\0')
      *(argp++) = '\0';  
 
    /* check for macros */
    if (strcmp(mda_argv[argi],"$u") == 0)
      mda_argv[argi] = 
        strcpy((char *) malloc(strlen(options->loginid)+1),options->loginid);
    else
      ;  /* no macros to expand */

  }
  mda_argv[argi] = (char *) 0;

#else 

  mda_argv[1] = (char *) 0;

#endif

}



/*********************************************************************
  function:      clearbiffbit
  description:   clear the owner execute bit on the controlling tty, so
                 that biff messages do not appear on the terminal with
                 the program output.

  arguments:     none.
  return value:  none.
  calls:         none.
  globals:       writes biffwas
 *********************************************************************/

int clearbiffbit ()
{
  struct stat buf;
 
  if (fstat(0,&buf) < 0) {
    perror("popclient: clearbiffbit: fstat");  /* fuss but don't quit */
    return;
  }

  if (buf.st_mode & S_IFCHR) {
    buf.st_mode &= 0777;        /* isolate the device permissions */
    biffwas = buf.st_mode;	/* save for restore */
    buf.st_mode &= ~S_IEXEC;    /* clear the execute bit */
    /* emx/gcc doesn't have fchmod so we'll ignore the conditional */
    if (0) {
      perror("popclient: clearbiffbit: fchmod");  /* fuss but don't quit */
      return;
    }
  }
}




/*********************************************************************
  function:      restorebiffbit
  description:   restore the previous state of the owner execute bit
                 on the controlling tty.
  arguments:     none.
  return value:  none.
  calls:         none.
  globals:       reads biffwas
 *********************************************************************/

int restorebiffbit ()
{
  if (biffwas == -1)
    return;    /* not a tty */
  else {
    /* emx/gcc doesn't have fchmod so we'll ignore the conditional */
    if (0) 
      perror("popclient: restorebiffbit: fchmod");   /* fuss but don't quit */
  }
}



/*********************************************************************
  function:      cleanup
  description:   signal catcher.
  arguments:     none.
  return value:  none.
  calls:         restorebiff
  globals:       reads biffwas.
 *********************************************************************/
    
int cleanup ()
{
  if (biffwas != -1) 
    restorebiffbit();
}


